plugins/syntax/numeral.ml

```diff
@@ -128,13 +128,212 @@ let warn_deprecated_decimal =
+       List.map (fun r ->
+           let c = match r with
+             | GlobRef.VarRef v -> Constr.mkVar v
+             | GlobRef.ConstRef c -> Constr.mkConst c
+             | GlobRef.IndRef i -> Constr.mkInd i
+             | GlobRef.ConstructRef c -> Constr.mkConstruct c in
+           let l = prod_length (get_type env sigma c) - nb_params in
+           r, r, args @ List.init l (fun _ -> ToPostCopy))
+         pt_refs in
+     [|l|], pt_refs
+
+let elaborate_to_post_via env sigma ty_ind ty_name l =
+     let ty_ind = Constr.mkInd ty_ind in
+     let sigma, ty_name =
+       locate_global_sort_inductive_or_constant sigma ty_name in
+     (* retrieve constants and constructors mappings and their type *)
+     let l =
+       let read (c, i) =
+         let lc, li = c.CAst.loc, i.CAst.loc in
+         let gc, c = locate_global_constructor_inductive_or_constant c in
+         let gi, i =
+           let i = Smartlocate.global_constructor_with_alias i in
+           GlobRef.ConstructRef i, Constr.mkConstruct i in
+         let ic = Impargs.(select_stronger_impargs (implicits_of_global gc)) in
+         lc, gc, c, get_type env sigma c, ic, li, gi, i, get_type env sigma i in
+       List.map read l in
+     let module CSet = CSet.Make (Constr) in
+     let module CMap = CMap.Make (Constr) in
+     let rec last_prod c = match Constr.kind c with
+       | Constr.Prod (_, _, c) -> last_prod c
+       | _ -> c in
+     (* in presence of implicit arguments, ignore all parameters
+        this is heuristic and unfortunately not entirely satisfactory *)
+     let l =
+       let ignored_args =
+         List.fold_left (fun s (_,_,_,t,i,_,_,_,_) ->
+             if List.for_all ((=) None) i then s else
+               match Constr.kind (last_prod t) with
+               | Constr.App (t, _) -> CSet.add t s
+               | _ -> s)
+           CSet.empty l in
+       let rec ignore c = match Constr.kind c with
+         | Constr.App (t, a) when CSet.mem t ignored_args -> t
+         | _ -> Constr.map ignore c in
+       let rec remove_implicits tc ic = match Constr.kind tc, ic with
+         | Constr.Prod (_, b, c), Some _ :: ic -> remove_implicits c ic
+         | Constr.Prod (a, b, c), None :: ic ->
+            Constr.mkProd (a, b, remove_implicits c ic)
+         | _ -> tc in
+       let clean tc ic = ignore (remove_implicits tc ic) in
+       List.map
+         (fun (lc,gc,c,tc,ic,li,gi,i,ti) -> lc,gc,c,clean tc ic,ic,li,gi,i,ti)
+         l in
+     (* associate inductives to types, and check that this mapping is one to one
+        and maps [ty_ind] to [ty_name] *)
+     let ind2ty, ty2ind =
+       List.fold_left (fun (ind2ty, ty2ind) (locc, _, c, tc, _, loci, _, i, ti) ->
+           let add loc l l' m =
+             try
+               let l'' = CMap.find l m in
+               if not (Constr.equal l' l'') then
+                 remapping_error ?loc env sigma l l'' l';
+               m
+             with Not_found -> CMap.add l l' m in
```

Comment:
I'd prefer better names than `l`, `l'`, `l''`.  I'm having trouble making sense of them.